README.md
# Cet

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Total Downloads][ico-downloads]][link-downloads]

CET score inquiry 大学英语四六级成绩查询

## Install

Via Composer

``` bash
$ composer require xu42/cet
```

## Usage

Please see [debug.php](debug.php)

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

Tests unavailable.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please using the issue tracker.

## Credits

- [Xu42](https://github.com/xu42)
- [All Contributors](https://github.com/xu42/cet/contributors)

## License

The GPL2.0 License. Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/xu42/cet.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/xu42/cet.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/xu42/cet
[link-downloads]: https://packagist.org/packages/xu42/cet
[link-author]: https://github.com/xu42
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

## 2.0.0 - 2017-04-23

* 大版本更新, 添加异常抛出, 接口地址改变

## 1.0.0 - 2016-06-11

* First release


src/BizImpl/CetBizImpl.php
<?php

namespace Cn\Xu42\Cet\BizImpl;

use Cn\Xu42\Cet\Exception\ArgumentException;

class CetBizImpl
{
    public function query($name, $number)
    {
        if (empty($name) || empty($number)) throw new ArgumentException('姓名和准考证号不能为空');

        $url = 'http://www.chsi.com.cn/cet/query?zkzh=' . $number . '&xm=' . urlencode($name);

        $curlResponse = $this->curlRequest($url);

        return $this->reCet($curlResponse);
    }

    private function curlRequest($url)
    {
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_REFERER, 'http://www.chsi.com.cn/cet/');
        curl_setopt($ch, CURLOPT_TIMEOUT, 5);
        curl_setopt($ch, CURLOPT_BINARYTRANSFER, true);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:24.0) Gecko/20100101 Firefox/24.0');
        $curlResponse = curl_exec($ch);
        curl_close($ch);
        return $curlResponse;
    }

    private function reCet($content)
    {
        preg_match_all('/<table(.*?)<\/table>/s', $content, $matches);
        preg_match_all('/(>)(.*?)(<)/s', $matches[0][1], $matches);

        $search = ['<', '>', '：', chr(13) . chr(10)];
        $replaceRes = str_replace($search, '', $matches[0]);

        foreach ($replaceRes as $value) { //去除 数组value 前后的空格
            $arr[] = trim($value);
        }

        foreach (array_filter($arr) as $value) { // 数组key重新排序
            $cetScores[] = $value;
        }

        $scores = [
            'name' => (string)$cetScores[3],
            'school' => (string)$cetScores[7],
            'type' => (string)$cetScores[9],
            'written' => [
                'number' => (string)$cetScores[12],
                'score' => (int)$cetScores[16],
                'listening' => (int)$cetScores[20],
                'reading' => (int)$cetScores[24],
                'translation' => (int)$cetScores[26]
            ],
            'oral' => [
                'number' => (string)$cetScores[29],
                'score' => (string)$cetScores[33]
            ]
        ];

        return $scores;
    }

}


src/CetScore.php
<?php

namespace Xu42\Cet;

class CetScore
{
    /**
     * 查询四六级成绩
     * @param string $number 准考证号
     * @param string $name 姓名
     * @return array
     */
    public function get($number, $name)
    {
        if (is_null($number) || is_null($name)) {
            return null;
        }
        $url = $this->setUrl($number, $name);
        $webPage = $this->getWebPage($url);
        return $this->cetScore($webPage);
    }


    private function setUrl($number, $name)
    {
        return 'http://www.chsi.com.cn/cet/query?zkzh=' . $number . '&xm=' . urlencode($name);
    }

    /**
     * @param string $url 请求的网址
     * @return string 包含成绩信息的网页源码
     */
    private function getWebPage($url)
    {
        if (is_null($url)) {
            return null;
        }

        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_REFERER, 'http://www.chsi.com.cn/cet/');
        curl_setopt($ch, CURLOPT_BINARYTRANSFER, true);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:24.0) Gecko/20100101 Firefox/24.0');
        $webPage = curl_exec($ch);
        curl_close($ch);
        return $webPage;
    }

    /**
     * @param string $webPage 包含成绩信息的网页源码
     * @return array 个人和四六级成绩等信息的数组（该数组尚不完美）
     */
    private function cetScore($webPage)
    {
        if (is_null($webPage)) {
            return null;
        }

        preg_match_all('/<table(.|\s)*?<\/table>/', $webPage, $matches);
        preg_match_all('/(>)(.|\s)*?(<)/', $matches[0][1], $matches);

        $search = ['<', '>', '：', chr(13) . chr(10)];
        $result = str_replace($search, '', $matches[0]);

        foreach ($result as $value) { //去除 数组value 前后的空格
            $content[] = trim($value);
        }

        foreach (array_filter($content) as $value) { // 数组key重新排序
            $cetScore[] = $value;
        }

        $score = [
            'name' => (string)$cetScore[3],
            'school' => (string)$cetScore[7],
            'type' => (string)$cetScore[9],
            'written' => [
                'number' => (string)$cetScore[12],
                'score' => (int)$cetScore[16],
                'listening' => (int)$cetScore[20],
                'reading' => (int)$cetScore[24],
                'translation' => (int)$cetScore[26]
            ],
            'oral' => [
                'number' => (string)$cetScore[29],
                'score' => (string)$cetScore[33]
            ]
        ];

        return $score;
    }

}
